### ISSUE TITLE
rabbitmq error "operation basic.publish caused a connection exception unexpected_frame: "expected content header for class 60, got non content header frame instead"

### ISSUE THREAD

**Issue by philippem** [syslog-ng/syslog-ng#3820]
# syslog-ng
## Version of syslog-ng

- `syslog-ng` in docker image `balabit/syslog-ng:latest` digest `9355a27dde82`
- in syslog-ng.conf `@version:3.31`

- `rabbitmq` in docker image `rabbitmq:3-management` digest `d5011653c255`, vanilla config

## Platform

- docker 20.10.9 running CentOS Linux release 7.9.2009 (Core) with kernel 3.10.0-1160.36.2.el7.x86_64

## Debug bundle
Will attempt to create a debug bundle.
# Issue
## Failure

Scenario:
- 5 linux VMs use syslog over ssh to send messages to syslog-ng container, ~5 messages/second average
- syslog uses ampq and elastic plugins
- occasionally, rabbitmq will refuse messages with error
```
2021-10-23 15:15:59.210190+00:00 [erro] <0.875.0> Error on AMQP connection <0.875.0> (172.16.238.20:58708 -> 172.16.238.21:5672, vhost: '/', user: 'guest', state: running), channel 1:
2021-10-23 15:15:59.210190+00:00 [erro] <0.875.0>  operation basic.publish caused a connection exception unexpected_frame: "expected content header for class 60, got non content header frame instead"
```
- syslog ng reports
```
...
[2021-10-23T15:14:58.817048] Log statistics; processed='global(payload_reallocs)=642027', processed='global(sdata_updates)=0', queued='global(scratch_buffers_bytes)=0', processed='global(msg_clones)=642027', processed='src.internal(s_local#0)=53', stamp='src.internal(s_local#0)=1635000902', dropped='dst.http(d_elastic#0,http,http://elasticsearch:9200/_bulk)=0', processed='dst.http(d_elastic#0,http,http://elasticsearch:9200/_bulk)=1398305', queued='dst.http(d_elastic#0,http,http://elasticsearch:9200/_bulk)=0', written='dst.http(d_elastic#0,http,http://elasticsearch:9200/_bulk)=1398305', processed='source(s_local)=53', dropped='dst.amqp(d_rabbitmq#0,amqp,/,rabbitmq,5672,syslog-ng,headers)=0', processed='dst.amqp(d_rabbitmq#0,amqp,/,rabbitmq,5672,syslog-ng,headers)=1398305', queued='dst.amqp(d_rabbitmq#0,amqp,/,rabbitmq,5672,syslog-ng,headers)=0', written='dst.amqp(d_rabbitmq#0,amqp,/,rabbitmq,5672,syslog-ng,headers)=1398305', processed='source(s_network)=1398313', processed='global(internal_queue_length)=0', processed='destination(d_elastic)=1398305', connections='src.syslog(s_network,afsocket_sd.(stream,AF_INET(0.0.0.0:601)))=5', dropped='global(internal_source)=0', queued='global(internal_source)=0', processed='center(received)=1398366', processed='destination(d_in)=1398305', processed='destination(d_rabbitmq)=1398305', queued='global(scratch_buffers_count)=568', processed='center(queued)=4194915'
[2021-10-23T15:15:59.208558] Network error while inserting into AMQP server; driver='d_rabbitmq#0', error='(unknown error)', time_reopen='60'
[2021-10-23T15:15:59.208650] Error occurred while trying to send a message, trying again; driver='d_rabbitmq#0', location='/etc/syslog-ng/syslog-ng.conf:30:5', worker_index='0', retries='1', time_reopen='60', batch_size='1'
[2021-10-23T15:16:59.264906] Network error while inserting into AMQP server; driver='d_rabbitmq#0', error='(unknown error)', time_reopen='60'
[2021-10-23T15:16:59.264989] Error occurred while trying to send a message, trying again; driver='d_rabbitmq#0', location='/etc/syslog-ng/syslog-ng.conf:30:5', worker_index='0', retries='2', time_reopen='60', batch_size='1'
...

```

- the error occurs occasionally, and it's not clear how to reproduce it

## Configuration

```
@version:3.31

@include "scl.conf"

options {
    keep-hostname(yes);
};

source s_network {
    syslog(
        ip(0.0.0.0)
        port(601)
        transport("tcp")
        max-connections(8192)
    );
};

destination d_elastic {
    elasticsearch-http(
        url("http://elasticsearch:9200/_bulk")
        index("syslog-ng_${YEAR}.${MONTH}.${WEEK}")
        type("")
        template("$(format-json --scope rfc5424 --scope nv-pairs --exclude DATE --key ISODATE)")
        time-zone("UTC")
    );
};


destination d_rabbitmq {
    amqp(
        vhost("/")
        host("rabbitmq")
        port(5672)
        username("guest")
        password("guest")
        exchange("syslog-ng")
        exchange-type("headers")
        routing-key("")
        body("$(format-json --scope rfc5424 --scope nv-pairs --exclude DATE --key ISODATE)")
        persistent(yes)
        exchange-declare(yes)
        value-pairs(
            scope("selected-macros" "nv-pairs" "sdata")
        )
    );
};

                
destination d_in {
    file("/var/log/syslog-ng/in.log"); 
};

parser p_json {
    json-parser (prefix("msg."));
};

log {
    source(s_network);
    if (message('^{')) {
       parser(p_json);
    } else {
    };
    destination(d_elastic);
    destination(d_rabbitmq);
    destination(d_in);
    flags(flow-control);
};

source s_local { internal(); };

#destination d_file { file("/var/log/syslog-ng/syslog-ng.log"); };

log {
    source(s_local);
#    destination(d_file);
};

```

**Comment by alltilla** [comment]
Hi @philippem 

Sorry for the late answer.
Thank you for your very detailed bug report!

I managed to reproduce an issue similar to this with a little config modification. I think the problem is with the size of the message.

My reproduction steps were the following:
  1. I added the `frame-size(8192)` option to the `amqp()` destination.
  2. I sent a message to syslog-ng with a size of ~8192.
  3. The `amqp()` destination tried to send the message, but it failed to do so, and I received the same error as you.
 
After a bit of debugging and error reporting fix I found out that amqp failed with the following error: "table too large for buffer".

---

Okay, but you didn't set `frame-size()` and its default value is 128K, which is pretty big, also by default syslog-ng truncates incoming messages larger than 64K so in theory every message must have been smaller than the allowed frame size. So how could have this happened to you?

The answer to that is most probably the json parser. Let's say you received a 64K sized, json formatted message. When the json parser parses a field, it stores it to a new name-value pair. This means, that if your message had one 64K large field, when it gets parsed, the total data stored is 128K, and it will be the data size sent to rabbitmq.

I managed to reproduce the issue with a 8K frame-size and a 4K json formatted message.

---

What about a fix?

Well, the rabbitmq client lib we use cannot handle larger messages than `frame-size()`, but you can set it to a larger value in your config, e.g.: `frame-size(1048576)` for 1M.

We have an error reporting bug, as you saw in your syslog-ng logs: (`error='(unknown error)'`). I will fix this.

Also, we could handle the too large messages differently, because they won't ever be sent successfully. In this case we should drop the message and continue with the next one. I will implement this, too.

Cheers,
Attila

**Comment by philippem** [comment]
Thank you for the fix and response, I'll test the `frame-size(1048576)` config setting.

**Comment by gaborznagy** [comment]
The issue is automatically closed, because #3869 is merged.
@philippem feel free to reopen it, in case you find some other issues.

**Comment by philippem** [comment]
I tried the suggested workaround first with `frame-size(1048576)`, then with `frame-size(10485760)`:

```
destination d_rabbitmq {
    amqp(
        frame-size(10485760)
...
```
in both cases, I saw the same errors. From `syslog-ng`:
```

[2022-01-06T16:44:09.439098] Network error while inserting into AMQP server; driver='d_rabbitmq#0', error='(unknown error)', time_reopen='60'
[2022-01-06T16:44:09.439497] Error occurred while trying to send a message, trying again; driver='d_rabbitmq#0', location='/etc/syslog-ng/syslog-ng.conf:30:5', worker_index='0', retries='1', time_reopen='60', batch_size='1'
[2022-01-06T16:45:09.484234] Network error while inserting into AMQP server; driver='d_rabbitmq#0', error='(unknown error)', time_reopen='60'
[2022-01-06T16:45:09.484391] Error occurred while trying to send a message, trying again; driver='d_rabbitmq#0', location='/etc/syslog-ng/syslog-ng.conf:30:5', worker_index='0', retries='2', time_reopen='60', batch_size='1'
[2022-01-06T16:46:09.545876] Network error while inserting into AMQP server; driver='d_rabbitmq#0', error='(unknown error)', time_reopen='60'
[2022-01-06T16:46:09.546033] Multiple failures while sending message(s) to destination, message(s) dropped; driver='d_rabbitmq#0', location='/etc/syslog-ng/syslog-ng.conf:30:5', worker_index='0', retries='3', batch_size='1'
```

on the rabbitmq side:

```
2022-01-06 16:44:09.439620+00:00 [erro] <0.28422.10> Error on AMQP connection <0.28422.10> (172.16.238.20:54070 -> 172.16.238.21:5672, vhost: '/', user: 'guest', state: running), channel 1:
2022-01-06 16:44:09.439620+00:00 [erro] <0.28422.10>  operation basic.publish caused a connection exception unexpected_frame: "expected content header for class 60, got non content header frame instead"
2022-01-06 16:44:09.440432+00:00 [info] <0.28422.10> closing AMQP connection <0.28422.10> (172.16.238.20:54070 -> 172.16.238.21:5672, vhost: '/', user: 'guest')

```

This is using the `balabit/syslog-ng` docker container with image id 123b82e5cba8, inside it's running syslog-ng 3.5.1-1:

```
root@51f2126902ad:/# dpkg -l | grep syslog
ii  syslog-ng                         3.35.1-1                       all          Enhanced system logging daemon (metapackage)
ii  syslog-ng-core                    3.35.1-1                       amd64        Enhanced system logging daemon (core)
ii  syslog-ng-mod-add-contextual-data 3.35.1-1                       amd64        Enhanced system logging daemon (add-contextual-data plugin)
ii  syslog-ng-mod-amqp                3.35.1-1                       amd64        Enhanced system logging daemon (AMQP plugin)
```

I'll try your fix (and look into rebuilding a docker image with it).

**Comment by philippem** [comment]
Could you prepare binaries of latest tip so I can test?

**Comment by MrAnno** [comment]
You can download a CentOS 7 package from here: https://github.com/syslog-ng/syslog-ng/actions/runs/1662582645

**Comment by alltilla** [comment]
I have been playing with this a bit, and there is a server side option that also needs to be set to change the maximal frame size.

Steps:
  * add `frame-size(1048576)` to `syslog-ng.conf`
  * add `frame_max = 1048576` to your rabbitmq config file

With these, I managed to send messages up to 1 MB each.

**Comment by philippem** [comment]
I'll test with `frame_max = 1048576` in the rabbitmq's `/etc/rabbitmq/rabbitmq.conf`. 
However, in our environment, we have rarely seen messages exceed 50kb.

**Comment by philippem** [comment]
@alltilla When will these changes make it into a new release of syslog-ng?

**Comment by alltilla** [comment]
Yeah, in theory the `frame_max` option should not limit the received message size, as the client should split the message into multiple frames, if it does not fit into one. However it seems like that the lib we are using (`librabbitmq-c`) does not support this splitting. Maybe I overlooked something about enabling it, but I was not able to find anything about it.

3.36.1 is planned to be released in mid-March, but I cannot give any guarantees.

**Comment by MrAnno** [comment]
Addressed by #3869 .
